services/scheduler/scheduler.go

```diff
@@ -0,0 +1,316 @@
+package scheduler
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/AlekSi/pointer"
+
+	"github.com/percona/pmm-managed/models"
+
+	"github.com/go-co-op/gocron"
+	"github.com/sirupsen/logrus"
+	"gopkg.in/reform.v1"
+)
+
+// Service is responsive for executing tasks and storing them to DB.
+type Service struct {
+	db        *reform.DB
+	scheduler *gocron.Scheduler
+	l         *logrus.Entry
+	tasks     map[string]context.CancelFunc
+	taskMx    sync.RWMutex
+	jobsMx    sync.Mutex
+}
+
+// New creates new scheduler service.
+func New(db *reform.DB) *Service {
+	scheduler := gocron.NewScheduler(time.UTC)
+	scheduler.TagsUnique()
+	return &Service{
+		db:        db,
+		scheduler: scheduler,
+		tasks:     make(map[string]context.CancelFunc),
+		l:         logrus.WithField("component", "scheduler"),
+	}
+}
+
+// Run loads tasks from DB and starts scheduler.
+func (s *Service) Run(ctx context.Context) {
+	if err := s.loadFromDB(); err != nil {
+		s.l.Warn(err)
+	}
+	s.scheduler.StartAsync()
+	<-ctx.Done()
+	s.scheduler.Stop()
+}
+
+// Add adds task to scheduler and save it to DB.
+func (s *Service) Add(task Task, cronExpr string, startAt time.Time, retry uint, retryInterval time.Duration) (*models.ScheduledTask, error) {
+	var scheduledTask *models.ScheduledTask
+	var err error
+	s.jobsMx.Lock()
+	defer s.jobsMx.Unlock()
+
+	err = s.db.InTransaction(func(tx *reform.TX) error {
+		scheduledTask, err = models.CreateScheduledTask(tx.Querier, models.CreateScheduledTaskParams{
+			CronExpression: cronExpr,
+			StartAt:        startAt,
+			Type:           task.Type(),
+			Data:           task.Data(),
+			Retries:        retry,
+			RetryInterval:  retryInterval,
+			Disabled:       false,
+		})
+		if err != nil {
+			return err
+		}
+
+		id := scheduledTask.ID
+		task.SetID(id)
+		fn := s.wrapTask(task, id, int(retry), retryInterval)
+
+		j := s.scheduler.Cron(cronExpr).SingletonMode()
+		if !startAt.IsZero() {
+			j = j.StartAt(startAt)
+		}
+		scheduleJob, err := j.Tag(id).Do(fn)
+		if err != nil {
+			return err
+		}
+
+		scheduledTask, err = models.ChangeScheduledTask(tx.Querier, id, models.ChangeScheduledTaskParams{
+			NextRun: scheduleJob.NextRun(),
+			LastRun: scheduleJob.LastRun(),
+		})
+		if err != nil {
+			s.l.WithField("id", id).Errorf("failed to set next run for new created task")
```

Comment:
can we return an error here and discard the scheduled job?

Reply:
do you think it's better to discard it, just because of next run wasn't saved to DB?

Reply:
I agree with Max, if we failed to save it here then it means something wrong, and better to return the error to the user.